File: leetcode/src/main/java/leetcode/editor/cn/蛇梯棋snakesAndLadders.java
/**
 * @Title: 909.蛇梯棋
 * @TitleSlug: snakesAndLadders
 * @Author: Neo
 * @Date: 2024-03-20 17:35:12
 */
package leetcode.editor.cn;

import java.util.ArrayDeque;
import java.util.Deque;

public class 蛇梯棋snakesAndLadders {
    public static void main(String[] args) {
        // 测试代码
        // Solution solution = new 蛇梯棋snakesAndLadders().new Solution();
    }
    
    // leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        int n;
        
        public int snakesAndLadders(int[][] board) {
            this.n = board.length;
            int ans = 0;
            Deque<Integer> queue = new ArrayDeque<>();
            queue.add(1);
            boolean[] visited = new boolean[n * n + 1];
            visited[1] = true;
            while (!queue.isEmpty()) {
                for (int k = queue.size(); k > 0; k--) {
                    int cur = queue.poll();
                    for (int x = cur + 1; x <= Math.min(cur + 6, n * n); x++) {
                        int dest = x;
                        int[] pos = getPos(x);
                        if (board[pos[0]][pos[1]] != -1) {
                            dest = board[pos[0]][pos[1]];
                        }
                        if (visited[dest]) {
                            continue;
                        }
                        if (dest == n * n) {
                            return ans + 1;
                        }
                        visited[dest] = true;
                        queue.add(dest);
                    }
                }
                ans++;
            }
            return -1;
        }
        
        private int[] getPos(int num) {
            num--;
            int x = n - 1 - num / n;
            int y = num % n;
            if (x % 2 == n % 2) {
                y = n - 1 - y;
            }
            return new int[]{x, y};
        }
        
        private int getNum(int x, int y) {
            if (x % 2 == n % 2) {
                y = n - 1 - y;
            }
            return n * (n - 1 - x) + y + 1;
        }
    }
// leetcode submit region end(Prohibit modification and deletion)

}
       

File: leetcode/src/main/python/leetcode/editor/cn/蛇梯棋snakesAndLadders.py
"""
Title: 909.蛇梯棋
Title Slug: snakesAndLadders
Author: Neo
Date: 2025-06-06 19:25:31
"""
from collections import deque
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def snakesAndLadders(self, board: List[List[int]]) -> int:
        n = len(board)

        def get_pos(num: int) -> (int, int):
            num -= 1
            i = n - 1 - num // n
            j = num % n
            if i & 1 == n & 1:
                j = n - 1 - j
            return i, j

        seen = [False] * (n * n + 1)
        seen[1] = True
        q = deque()
        q.append(1)
        ans = 0
        while q:
            for _ in range(len(q)):
                x = q.popleft()
                for y in range(x + 1, min(x + 6, n * n) + 1):
                    if seen[y]:
                        continue
                    seen[y] = True
                    dest = y
                    i, j = get_pos(y)
                    if board[i][j] != -1:
                        dest = board[i][j]
                    if dest == n * n:
                        return ans + 1
                    q.append(dest)
            ans += 1
        return -1
# leetcode submit region end(Prohibit modification and deletion)
